### PR TITLE
Add Azure Cloud Shell support in Rider plugin

### PR DESCRIPTION
**Pull request by maartenba**
Add Azure Cloud Shell support in Rider plugin. Currently supports:
* Selecting subscription to run cloud shell in
* Open a terminal to the cloud shell and work with it
* Support for remote disconnect due to idle time
* Support for resizing terminal
* Upload to cloud shell action
* Intercept `download` command and open browser to fetch file from cloud shell

tada

Note to self: need to PR https://docs.microsoft.com/en-us/azure/cloud-shell/overview when plugin is released.

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/cloudshell/actions/StartAzureCloudShellAction.kt [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                        tokenCredentials)
+
+                val consoleUserSettingsResponse = retrofitClient.userSettings().execute()
+                if (consoleUserSettingsResponse.isSuccessful) {
+                    indicator.fraction = 0.35
+                    indicator.text = "Provisioning cloud shell..."
+
+                    // We can provision a shell!
+                    val provisionParameters = CloudConsoleProvisionParameters(
+                            CloudConsoleProvisionParameters.Properties(consoleUserSettingsResponse.body()!!.properties!!.preferredOsType!!))
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/cloudshell/actions/StartAzureCloudShellAction.kt]: Should not we check values for all those force unwrapping? Here and below.

**Review by sdubov — COMMENTED** [review]
Overall, I like the change. I'll wait till other guys make their review.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/cloudshell/actions/StartAzureCloudShellAction.kt (hunk 1 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/cloudshell/actions/StartAzureCloudShellAction.kt]: Good idea. Updated.

**Comment by IvanPashchenko** [comment]
I'd like to have this a proper terminal rather that run configuration. How about we do it that way and in `183`?

**Comment by maartenba** [comment]
Re-using what is in the terminal plugin right now ([`CloudTerminalRunner`](https://github.com/JetBrains/intellij-community/blob/master/plugins/terminal/src/org/jetbrains/plugins/terminal/cloud/CloudTerminalRunner.java)). Is there another (better) way to hook up the in/out streams I have into the terminal?

**Comment by sdubov** [comment]
I think they might use run configuration intentionally. Otherwise, you have to keep connection open while you are using Rider. With run config you limit the session while configuration is running.

**Comment by IvanPashchenko** [comment]
> Otherwise, you have to keep connection open while you are using Rider.

Well, that would be convenient. When I worked with Azure, I've always had an open Azure Powershell which I switched to pretty often. Having to manually run a config instead of just having a terminal doesn't sound so cool.

**Comment by sdubov** [comment]
I fully agree that it would be much convenient to use a separate terminal. I doubt about keeping connection alive all the time. Imagine you install the Azure plugin and work with it for some time, but you usually run Rider for your regular Azure-unrelated development and might be even with no internet connection. Then you need to have some mechanism to re-establish connection (that might be not that bad if we don't ping user about missing connection all the time). Maybe setup connection when you open a terminal, and close it after some time if user don't use it. TDB...

**Comment by maartenba** [comment]
Azure itself closes connection after +/- 20 min idle time.

Updated sources, now displaying in terminal tool window: **with a downside, see below**
snim

**DOWNSIDE**: One downside is that it always opens a local terminal as well.

The local terminal comes from [here](https://github.com/JetBrains/intellij-community/blob/master/plugins/terminal/src/org/jetbrains/plugins/terminal/TerminalView.java#L77), where in an empty terminal window it always creates a local session [here](https://github.com/JetBrains/intellij-community/blob/master/plugins/terminal/src/org/jetbrains/plugins/terminal/TerminalView.java#L125).

So: prefer as run configuration or using the terminal window?

**Comment by maartenba** [comment]
No way around opening local shell as well, but at least now got it to focus on the Azure one all the time :-) (with a dirty `Thread.sleep()` to give the local terminal a few ms to start)

vnext

**Comment by maartenba** [comment]
Added:
* Support for resizing terminal
* Upload to cloud shell action

tada

Note to self: need to PR https://docs.microsoft.com/en-us/azure/cloud-shell/overview when plugin is released.

**Comment by vlomshakov** [comment]
Hey @maartenba ,
If I understand right, `createNewSession` method requests terminal tool window  activation and create a new session into callback.
`UIUtil.invokeLaterIfNeeded(() -> TerminalView.getInstance(project).createNewSession(project, runner))`
Also you can ask Dmitry Trofimov.

**Comment by maartenba** [comment]
Thanks, updated!

@sdubov Any other reviews needed?

**Comment by sdubov** [comment]
^^^ @IvanPashchenko

**Comment by IvanPashchenko** [comment]
nope, I think we can merge now :)

**Comment by maartenba** [comment]
Merged in `183` to make merge of PR easier. Good to go, @sdubov.

**Pull request merged**